spec/index.html

```diff
@@ -1492,3 +1502,413 @@ <h1>SharedArray ( ..._values_ )</h1>
+      <h1>Atomics.Condition.waitFor ( _cv_, _mutexUnlockToken_, _timeout_ [ , _predicate_ ] )</h1>
+      <p>If _predicate_ is *undefined*, this function atomically unlocks _mutexUnlockToken_ and puts the surrounding agent in a wait queue and suspends it until the condition variable is notified or until the wait times out, returning *true* for the former and *false* for the latter.</p>
+      <p>If a _predicate_ is passed and calling it returns *false*, this function atomically unlocks _mutexUnlockToken_ and puts the surrounding agent in a wait queue and suspends it until _predicate_ returns *true*, or until the wait times out. The return value of the final call to _predicate_ is returned.</p>
+      <p>It performs the following steps when called:</p>
+      <emu-alg>
+        1. Perform ? RequireInternalSlot(_cv_, [[ConditionWaiterList]]).
+        1. Perform ? RequireInternalSlot(_mutexUnlockToken_, [[LockedMutex]]).
+        1. Let _mutex_ be _mutexUnlockToken_.[[LockedMutex]].
+        1. If _mutex_ is ~empty~, throw a *TypeError* exception.
+        1. If _timeout_ is not a Number, throw a *TypeError* exception.
+        1. If _timeout_ is either *NaN* or *+∞*<sub>𝔽</sub>, let _tMillis_ be +∞; else if _timeout_ is *-∞*<sub>𝔽</sub>, let _tMillis_ be 0; else let _tMillis_ be max(ℝ(_timeout_), 0).
+        1. If _predicate_ is not *undefined* and IsCallable(_predicate_) is *false*, throw a *TypeError* exception.
+        1. If AgentCanSuspend() is *false*, throw a *TypeError* exception.
+        1. Let _thisAgent_ be AgentSignifier().
+        1. Let _WL_ be _cv_.[[ConditionWaiterList]].
+        1. Let _satisfied_ be *false*.
+        1. Repeat,
+          1. If _predicate_ is not *undefined*, then
+            1. Set _satisfied_ to ToBoolean(? Call(_predicate_, *undefined*)).
+          1. If _satisfied_ is *true*, return *true*.
+          1. If _tMillis_ &le; 0, return _satisfied_.
+          1. Let _timeBeforeWait_ be the time value (UTC) identifying the current time.
+          1. Let _additionalTimeout_ be an implementation-defined non-negative mathematical value.
+          1. Let _timeoutTime_ be ℝ(_timeBeforeWait_) + _tMillis_ + _additionalTimeout_.
+          1. NOTE: When _tMillis_ is +∞, _timeoutTime_ is also +∞.
+          1. Perform EnterCriticalSection(_WL_).
+          1. Let _waiterRecord_ be a new Waiter Record { [[AgentSignifier]]: _thisAgent_, [[PromiseCapability]]: ~blocking~, [[TimeoutTime]]: _timeoutTime_, [[Result]]: *"ok"* }.
+          1. Perform AddWaiter(_WL_, _waiterRecord_).
+          1. Perform UnlockMutex(_mutex_).
+          1. Perform SuspendThisAgent(_WL_, _waiterRecord_).
+          1. Perform LeaveCriticalSection(_WL_).
+          1. Perform LockMutex(_mutex_, +∞).
+          1. Let _timeAfterWait_ be the time value (UTC) identifying the current time.
+          1. Let _timeElapsed_ be _timeAfterWait_ - _timeBeforeWait_.
+          1. Set _tMillis_ to _tMillis_ - _timeElapsed_.
```

Comment:
Is there a value in this approach rather than just calculating the absolute timeout prior to entering the loop and determining the remaining timeout prior to waiting? Recalculating the remaining time here happens before executing user code, which could result in the condition waiting longer than it should.

i.e., if we invoke `waitFor(cv, token, 1000, pred)` and `pred` takes 100ms per each execution and is executed 10 times, then we actually could potentially end up waiting ~2000ms. User code should eat up part of the timeout.

In C++, `wait_for(lock, rel_time, pred)` is equivalent to `wait_until(lock, std::chrono::steady_clock::now + rel_time, std::move(pred))`, so user code executing in the predicate would also eat up part of the timeout.

Reply:
Yeah, that sounds like a good idea to include the predicate() time.